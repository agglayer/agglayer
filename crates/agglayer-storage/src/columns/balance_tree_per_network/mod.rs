use super::{ColumnSchema, BALANCE_TREE_PER_NETWORK_CF};

#[cfg(test)]
mod tests;

/// Column family for the balance tree per network.
///
/// ## Column definition
///
/// | key                               | value                                   |
/// | ---                               | --                                      |
/// | (`NetworkId`, `SmtKeyType::Root`) | (`hash(root.left)`, `hash(root.right)`) |
/// | (`NetworkId`, `hash(node)`)       | (`hash(node.left)`, `hash(node.right)`) |
/// | (`NetworkId`, `hash(leaf)`)       | (`hash(leaf)`)                          |
pub struct BalanceTreePerNetworkColumn;

impl ColumnSchema for BalanceTreePerNetworkColumn {
    type Key = crate::types::SmtKey;
    type Value = crate::types::SmtValue;

    const COLUMN_FAMILY_NAME: &'static str = BALANCE_TREE_PER_NETWORK_CF;
}
