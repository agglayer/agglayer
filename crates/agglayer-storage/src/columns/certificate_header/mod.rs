use agglayer_types::CertificateHeader;

use super::{ColumnSchema, CERTIFICATE_HEADER_CF};

#[cfg(test)]
mod tests;

/// Column family for the certificate headers.
///
/// ## Column definition
///
/// | key               | value                |
/// | --                | --                   |
/// | `CertificateId`   |  `CertificateHeader` |
pub struct CertificateHeaderColumn;

pub type Key = agglayer_types::CertificateId;
pub type Value = CertificateHeader;

impl ColumnSchema for CertificateHeaderColumn {
    type Key = Key;
    type Value = Value;

    const COLUMN_FAMILY_NAME: &'static str = CERTIFICATE_HEADER_CF;
}
