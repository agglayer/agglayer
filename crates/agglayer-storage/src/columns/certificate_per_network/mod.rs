use agglayer_types::CertificateId;
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, CERTIFICATE_PER_NETWORK_CF};

#[cfg(test)]
mod tests;

/// Column family for the certificate per network per height.
///
/// ## Column definition
///
/// | key                     | value           |
/// | --                      | --              |
/// | (`NetworkId`, `Height`) | `CertificateId` |
pub struct CertificatePerNetworkColumn;

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Key {
    pub(crate) network_id: u32,
    pub(crate) height: u64,
}

impl Codec for Key {}

impl ColumnSchema for CertificatePerNetworkColumn {
    type Key = Key;
    type Value = CertificateId;

    const COLUMN_FAMILY_NAME: &'static str = CERTIFICATE_PER_NETWORK_CF;
}
