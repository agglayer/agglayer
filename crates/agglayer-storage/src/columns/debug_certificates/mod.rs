use agglayer_types::{Certificate, CertificateId};

use super::{ColumnSchema, DEBUG_CERTIFICATES_CF};

/// Column family containing the certificates received.
///
/// ## Column definition
///
/// | key             | value           |
/// | --              | --              |
/// | `CertificateId` | `Certificate`   |
pub(crate) struct DebugCertificatesColumn;

impl ColumnSchema for DebugCertificatesColumn {
    type Key = CertificateId;
    type Value = Certificate;

    const COLUMN_FAMILY_NAME: &'static str = DEBUG_CERTIFICATES_CF;
}
