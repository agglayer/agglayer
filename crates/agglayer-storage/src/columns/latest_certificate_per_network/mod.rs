use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF};
use crate::types::NetworkId;

#[cfg(test)]
mod tests;

/// Column family for the latest settled certificate per network.
/// The key is the network_id and the value is the certificateID,
/// the height and the epoch_number.
///
/// ## Column definition
///
/// | key         | value                                |
/// | --          | --                                   |
/// | `NetworkId` | (`Proof`, `CertificateId`, `Height`) |
pub struct LatestSettledCertificatePerNetworkColumn;

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProvenCertificate(pub [u8; 32], pub u64, pub u64);

pub type Key = NetworkId;

impl Codec for ProvenCertificate {}

impl ColumnSchema for LatestSettledCertificatePerNetworkColumn {
    type Key = Key;
    type Value = ProvenCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF;
}
