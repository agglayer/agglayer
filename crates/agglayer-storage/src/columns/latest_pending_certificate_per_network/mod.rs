use agglayer_types::{CertificateId, Height, NetworkId};
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF};

/// Column family for the latest pending certificate per network.
/// The key is the network_id and the value is the certificateID and
/// the height.
///
/// ## Column definition
///
/// | key         | value                       |
/// | --          | --                          |
/// | `NetworkId` | (`CertificateId`, `Height`) |
pub struct LatestPendingCertificatePerNetworkColumn;

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PendingCertificate(pub CertificateId, pub Height);

pub type Key = NetworkId;

impl Codec for PendingCertificate {}

impl ColumnSchema for LatestPendingCertificatePerNetworkColumn {
    type Key = Key;
    type Value = PendingCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF;
}
