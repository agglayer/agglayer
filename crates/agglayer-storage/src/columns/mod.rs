use bincode::{DefaultOptions, Options};
use serde::{de::DeserializeOwned, Serialize};

use crate::error::Error;

pub fn default_bincode_options() -> impl bincode::Options {
    DefaultOptions::new()
        .with_big_endian()
        .with_fixint_encoding()
}

// State related CFs
pub const CERTIFICATE_PER_NETWORK_CF: &str = "certificate_per_network_cf";
pub const NULLIFIER_TREE_PER_NETWORK_CF: &str = "nullifier_tree_per_network_cf";
pub const BALANCE_TREE_PER_NETWORK_CF: &str = "balance_tree_per_network_cf";
pub const LOCAL_EXIT_TREE_PER_NETWORK_CF: &str = "local_exit_tree_per_network_cf";

// Metadata CFs
pub const CERTIFICATE_HEADER_CF: &str = "certificate_header_cf";
pub const LATEST_PROVEN_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_proven_certificate_per_network_cf";
pub const LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_settled_certificate_per_network_cf";
pub const LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_pending_certificate_per_network_cf";
pub const METADATA_CF: &str = "metadata_cf";

// epochs related CFs
pub const PER_EPOCH_CERTIFICATES_CF: &str = "per_epoch_certificates_cf";
pub const PER_EPOCH_METADATA_CF: &str = "per_epoch_metadata_cf";
pub const PER_EPOCH_PROOFS_CF: &str = "per_epoch_proofs_cf";
pub const PER_EPOCH_END_CHECKPOINT_CF: &str = "per_epoch_end_checkpoint_cf";
pub const PER_EPOCH_START_CHECKPOINT_CF: &str = "per_epoch_start_checkpoint_cf";
pub const PER_EPOCH_TRANSACTION_HASH_PER_CERTIFICATE_INDEX: &str =
    "per_epoch_transaction_hash_per_certificate_index";

// Pending related CFs
pub const PENDING_QUEUE_CF: &str = "pending_queue_cf";
pub const PROOF_PER_CERTIFICATE_CF: &str = "proof_per_certificate_cf";

// debug CFs
pub const DEBUG_CERTIFICATES_CF: &str = "debug_certificates";

pub trait Codec: Sized + Serialize + DeserializeOwned {
    fn encode(&self) -> Result<Vec<u8>, Error> {
        Ok(default_bincode_options().serialize(self)?)
    }

    fn decode(buf: &[u8]) -> Result<Self, Error> {
        Ok(default_bincode_options().deserialize(buf)?)
    }
}

pub trait ColumnSchema {
    type Key: Codec;
    type Value: Codec;

    const COLUMN_FAMILY_NAME: &'static str;
}

// State
pub(crate) mod balance_tree_per_network;
pub(crate) mod certificate_per_network;
pub(crate) mod local_exit_tree_per_network;
pub(crate) mod nullifier_tree_per_network;

// Pending
pub(crate) mod pending_queue;
pub(crate) mod proof_per_certificate;

// Metadata
pub(crate) mod certificate_header;
pub mod latest_pending_certificate_per_network;
pub mod latest_proven_certificate_per_network;
pub mod latest_settled_certificate_per_network;
pub(crate) mod metadata;

// Debug
pub(crate) mod debug_certificates;

// PerEpoch
pub mod epochs {
    pub(crate) mod certificates;
    pub mod end_checkpoint;
    pub(crate) mod metadata;
    pub(crate) mod proofs;
    pub(crate) mod start_checkpoint;
    mod transaction_hash_per_certificate_index;
}
