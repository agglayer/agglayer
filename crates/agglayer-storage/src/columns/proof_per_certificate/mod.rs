use agglayer_types::{CertificateId, Proof};

use super::{ColumnSchema, PROOF_PER_CERTIFICATE_CF};

#[cfg(test)]
mod tests;

/// Column family that returns the generated proof for one certificate.
///
/// ## Column definition
///
/// | key             | value   |
/// | --              | --      |
/// | `CertificateId` | `Proof` |
pub struct ProofPerCertificateColumn;

impl ColumnSchema for ProofPerCertificateColumn {
    type Key = CertificateId;
    type Value = Proof;

    const COLUMN_FAMILY_NAME: &'static str = PROOF_PER_CERTIFICATE_CF;
}
