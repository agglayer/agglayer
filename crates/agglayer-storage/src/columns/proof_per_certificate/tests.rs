use insta::assert_snapshot;
use serde::Deserialize;
use sp1_sdk::SP1ProofWithPublicValues;

use crate::columns::{
    proof_per_certificate::{CertificateId, Proof},
    Codec as _,
};

#[test]
fn can_parse_key() {
    let key: CertificateId = [1; 32].into();

    let encoded = key.encode().expect("Unable to encode key");

    let expected_key = CertificateId::decode(&encoded[..]).expect("Unable to decode key");

    assert_eq!(expected_key, key);

    assert_eq!(
        encoded[..32],
        [
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
            1, 1, 1,
        ]
    );
}

#[test]
fn can_parse_value() {
    let value = Proof::dummy();

    let encoded = value.encode().expect("Unable to encode value");

    let expected_value = Proof::decode(&encoded[..]).expect("Unable to decode value");

    assert!(matches!(expected_value, Proof::SP1(_)));
}

#[derive(Deserialize)]
struct SP1ProofWithPublicValuesV3 {
    pub proof: sp1_sdk::SP1Proof,
    #[allow(unused)]
    pub stdin: sp1_sdk::SP1Stdin,
    pub public_values: sp1_sdk::SP1PublicValues,
    pub sp1_version: String,
}

impl SP1ProofWithPublicValuesV3 {
    fn load(path: impl AsRef<std::path::Path>) -> Result<Self, Box<dyn std::error::Error>> {
        bincode::deserialize_from(std::fs::File::open(path).expect("failed to open file"))
            .map_err(Into::into)
    }
}

impl From<SP1ProofWithPublicValuesV3> for SP1ProofWithPublicValues {
    fn from(v3: SP1ProofWithPublicValuesV3) -> Self {
        Self {
            proof: v3.proof,
            public_values: v3.public_values,
            sp1_version: v3.sp1_version,
        }
    }
}

#[test]
fn non_regression_proof_encoding() {
    let path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("./src/columns/proof_per_certificate/fixtures/non_regression_proof_encoding.proof");

    let proof: SP1ProofWithPublicValues = SP1ProofWithPublicValuesV3::load(path).unwrap().into();
    assert_snapshot!("proof hex format", hex::encode(proof.bytes()));

    assert_snapshot!(
        "proof public input hex format",
        hex::encode(proof.public_values.as_slice())
    );
}
