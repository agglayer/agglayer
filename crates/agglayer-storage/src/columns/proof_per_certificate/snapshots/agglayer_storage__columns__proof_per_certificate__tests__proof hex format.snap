---
source: crates/agglayer-storage/src/columns/proof_per_certificate/tests.rs
expression: "hex::encode(proof.bytes())"
---
54bdcae32e0bdbbe5b39fc2bbfcb6d3439b1b248667158a0bf9b17bd65f3afed25a696990b932a913ad2945715b681036317daaf5f64b22aef7bdc94ad634b6d271fd345153d91898e3a6f996ef7b091ff94c8db64b49845979a601b9d06668783e3a4bb26d8425d738eb09a9d2f5a52297e17c4ced2df415fb7816199354cf39e65b17d1bedaa9bb41563bdd03e68be9dc8bee0ed669e0b91dccea1a5d14f8dcfc1ba3212f71bd5b96690645e753705043f88d5237c6abc448cdb7c6a1ab50346d1f1b5276bb3dafdb6b084d56f37b4cd58e2bb8d5bd195fc8bc27b9935a548f73225761e9fc4812fc0b27dd127be7ed1584293d733e228fc979d7d1fba9b41faf18dae1f2873c7d9cda2b04f6a7b62bbb46dd27eb4b209a1188ed2ad1cba7cd53b11f4127b8fee4cc3b141efaff958d3db771d698721fd3de82636d5123acec3d9ed580673b4989dd5f5481b0773ab22092fb5ef0b01e4fed4186be3e73fbf26a0dd5b07b4b7e0ea43a285e8a779c438e948b83ddc21c2c29c10cacf0b167ea6a41b06076e893b888a30440d5e07f57a8d07998961ac20611959836acc1865f3ad2d390eee9127057b90c5f3f0c209964128c4f4d863205391b8c1e25ce59389db684b1352e86a7c729a3d71dcf86b59d58bb8286f5c4dd28aa2f9c83f60fc4647d101266e632d292c0c357935f4526b83c731962fc2e8b099ce0f55f002f8c6b143f02d7d6a12d823d837f379c6aa88a764f5a80f72e0442dba39bfbd809d821063dd20bac1717d5dc66c56123d4956ccbaae47dc97ad428e0891863d29f4e5be47b42821dcf8e5f7f338f799f87babac7d3c6d236659931f5ea8069b43ed48b99de2242c0e5da4066c29567e26d1a231f4c973e1e066cd3bbe143a03800ba245d6e7001bbcdba4b27ceb069f3de8740b8500b9708513f6533808809c0a17af4c9181281ab251d273228b36beb853ee1ecfb051750768ae542a411ffde88dfc1361f3203325709a5b132df738086ea9cfa32a3838d45389d124f6777cb492b271f3fd1ebf26f56fb87ab5b30dcb9a7b5964ac088c32df5aebb426dce0d636c2b2d55513c02a4c5fc7ed013eeb66feb902ae2587312dec5874f98b69c53a64b15faf2f1c7e317b0fac6cf3b138f9eabaeda19c7af6fb9c64606ddb982cf278ff47be57115adc1a64640249bc26a61f2de19d4dcacfb3b168f5ab297608f4852fcdf257
